### DISCUSSION TITLE
Head / NextScript cannot be used as a JSX component

### DISCUSSION THREAD

**Discussion by Kinbaum** [vercel/next.js#17205] [Help]
# Bug report

## Describe the bug

@Timer Using typescript, yarn 2, and a monorepo, when I create a `_document.tsx` file, I see the following errors:

```
Type error: 'Head' cannot be used as a JSX component.
  Its instance type 'Head' is not a valid JSX element.
    Type 'Head' is missing the following properties from type 'ElementClass': setState, forceUpdate, props, state, refs

Type error: 'NextScript' cannot be used as a JSX component.
  Its instance type 'NextScript' is not a valid JSX element.
    Type 'Head' is missing the following properties from type 'ElementClass': setState, forceUpdate, props, state, refs
```

## To Reproduce

I'm not sure how to reproduce this error... I understand that this is needed, but maybe I can get some guidance on where to begin. I copied the `with-typescript` example and added my exact set up to it. After doing so, I do not see this error. So I'm a bit confused as to why it's showing up in my other repo when I add the `_document.tsx` file.

If I switch the file extension to `.js`, then my app builds no problem and my linter doesn't identify any issues with the file.

## Code Snippets

<details>
    <summary>_document.tsx</summary>

```tsx
import Document, { Html, Head, Main, NextScript } from 'next/document';

class MyDocument extends Document {

    render() {
        return (
            <Html>
                { /*
                    Type error: 'Head' cannot be used as a JSX component.
                    Its instance type 'Head' is not a valid JSX element.
                    Type 'Head' is missing the following properties from type 'ElementClass': setState, forceUpdate, props, state, refs
                */ }
                <Head>
                    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />
                    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png" />
                    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png" />
                    <link rel="manifest" href="/site.webmanifest" />
                </Head>
                <body>
                    <Main />
                    { /*
                        Type error: 'NextScript' cannot be used as a JSX component.
                        Its instance type 'NextScript' is not a valid JSX element.
                        Type 'NextScript' is missing the following properties from type 'ElementClass': setState, forceUpdate, props, state, refs
                    */ }
                    <NextScript />
                </body>
            </Html>
        );
    }

}

export default MyDocument;
```
</details>

<details>
    <summary>tsconfig.json</summary>

```json
{
    "compilerOptions": {
        "target": "es2019",
        "module": "esnext",
        "jsx": "preserve",
        "strict": true,
        "esModuleInterop": true,
        "skipLibCheck": true,
        "forceConsistentCasingInFileNames": true,
        "lib": [
            "dom",
            "dom.iterable",
            "esnext"
        ],
        "allowJs": true,
        "noEmit": true,
        "moduleResolution": "node",
        "resolveJsonModule": true,
        "isolatedModules": true,
        "types": [
            "node"
        ]
    },
    "exclude": [
        "node_modules"
    ],
    "include": [
        "**/*.ts",
        "**/*.tsx"
    ]
}
```
</details>

<details>
    <summary>console.log of the Head component</summary>

```
[class Head extends Component] {
  contextType: <ref *1> {
    '0': null,
    '1': {
      __NEXT_DATA__: [Object],
      buildManifest: [Object],
      docComponentsRendered: {},
      dangerousAsPath: '/',
      canonicalBase: '',
      ampPath: '',
      inAmpMode: false,
      isDevelopment: true,
      hybridAmp: false,
      dynamicImports: [],
      assetPrefix: '',
      headTags: [],
      unstable_runtimeJS: undefined,
      devOnlyCacheBusterQueryString: '?ts=1600444304200',
      html: '',
      head: [Array],
      styles: []
    },
    '$$typeof': Symbol(react.context),
    _calculateChangedBits: null,
    _currentValue: null,
    _currentValue2: null,
    _threadCount: 2,
    Provider: { '$$typeof': Symbol(react.provider), _context: [Circular *1] },
    Consumer: {
      '$$typeof': Symbol(react.context),
      _context: [Circular *1],
      _calculateChangedBits: null
    },
    _currentRenderer: null,
    _currentRenderer2: null,
    displayName: 'DocumentContext'
  },
  propTypes: {
    nonce: [Function: bound checkType] {
      isRequired: [Function: bound checkType]
    },
    crossOrigin: [Function: bound checkType] {
      isRequired: [Function: bound checkType]
    }
  }
}
```
</details>

## System information

- OS: macOS
- Version of Next.js: 9.5.4-canary.20
- Version of React: 17.0.0-rc.1 (happens on v16 too)
- Version of React DOM: 17.0.0-rc.1 (happens on v16 too)
- Version of Node.js: 14.9.0

**Comment by runewolf7** [comment]
Same error on my end. Seemed to begin once I upgraded to Next 9.5.3.

**Comment by myasul** [comment]
I'm experiencing the same error upon a new project via `npx create-next-app`.

**Comment by mattcarlotta** [comment]
Unable to replicate using Yarn v1. Not very experienced with Yarn v2, but that appears to be the problem...

Yarn v2 doesn't install a `node_modules` folder. Instead, it installs a `.yarn` folder that contains **zipped** assets (which can't be read by tsconfig). By default, tsconfig looks for types in `node_modules` which doesn't exist. So you'll have to do some tweaks to get it to work. I followed this [guide (scroll down to the Typescript section)](https://medium.com/swlh/getting-started-with-yarn-2-and-typescript-43321a3acdee) and it seems to be working:

<details>
<summary>tsconfig.json</summary>
<pre><code>
{
  "compilerOptions": {
    "allowJs": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "lib": ["dom", "esnext"],
    "module": "esnext",
    "moduleResolution": "node",
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "esnext"
  },
  "exclude": ["node_modules"],
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"]
}
</pre></code>
</details>

<details>
<summary>package.json</summary>
<pre><code>
{
  "name": "with-typescript",
  "version": "1.0.0",
  "scripts": {
    "dev": "next",
    "build": "next build",
    "start": "next start",
    "type-check": "tsc"
  },
  "dependencies": {
    "next": "^9.5.4",
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "devDependencies": {
    "@types/react": "^16.9.51",
    "@types/react-dom": "^16.9.8",
    "@yarnpkg/pnpify": "^2.3.2",
    "typescript": "4.0"
  },
  "license": "MIT"
}
</pre></code>
</details>

<img width="1680" alt="Screen Shot 2020-10-08 at 11 06 30 PM" src="https://user-images.githubusercontent.com/22607722/95550704-330f2980-09be-11eb-849d-c7b631f33abd.png">

**Comment by Kinbaum** [comment]
This has just started magically working for me

**Comment by EmmanoelCoutinho** [comment]
I forgot to install ts in my project, so after installing and restarting vscode the error is gone

**Reply by christiankaindl** [reply]
Yes! After restarting the dev server, Next.js suggested running `npm install --save-dev typescript @types/react @types/node`. That installed the necessary dependencies and also created a tsconfig.json file after the next `npm run dev`.

Fixed the issue for me.

**Reply by davidawad** [reply]
in my case I had outdated types, the yarn equivalent of this command worked for me.

**Comment by felipedeamonteiro** [comment]
I got the same error using Next 11.1.2

**Reply by nwabueze1** [reply]
where you able to resolve this

**Comment by MatheusCerqueiraDev** [comment]
Run yarn add typescript @types/react @types/node it should fix

**Reply by aafnnp** [reply]
> Run yarn add typescript @types/react @types/node it should fix

still got this problem 

```
"@types/node": "16.11.7",
"@types/react": "17.0.34",
"@types/react-dom": "^17.0.11",
"typescript": "4.4.4",
```

**Reply by BDMayhem** [reply]
> Seems like there is a bug in the current @types/react version (specifically v18.0.2), you can downgrade to 18.0.1 with `npm install --save-dev @types/react@18.0.1`

https://stackoverflow.com/questions/71843307/next-js-is-not-recognizing-types-react/71843538#71843538

**Reply by clemenspeters** [reply]
Thanks @MatheusCerqueiraDev 
`yarn add typescript @types/react @types/node -D` solved it for us (I'd recommend to use DevDependencies whenever possible).

**Comment by NishargShah** [comment]
add `resolutions` in `package.json` file, works for me.

```
"resolutions": {
    "@types/react": "^17.0.38"
}
```

**Comment by jameshfisher** [comment]
I finally tracked this down to `"allowSyntheticDefaultImports": false`. If I set it to `true`, the type error goes away. Bug reported at https://github.com/vercel/next.js/issues/40835. (I don't want to set this to `true`; synthetic default imports are gross.)

**Reply by lancejpollard** [reply]
Has no effect for me.

**Comment by PrimeObjects** [comment]
I did the steps below and fixed the issue

1. I found some of the packages were located in dependencies in package.json by mistake (this is done by npx create-next-app@latest --typescript). I relocated them properly below
    "dependencies": {
        "next": "13.1.2",
        "react": "18.2.0",
        "react-dom": "18.2.0"
    },
    "devDependencies": {
        "tailwindcss": "^3.2.4",
        "@next/font": "13.1.2",
        "@types/node": "18.11.18",
        "@types/react": "18.0.26",
        "@types/react-dom": "18.0.10",
        "eslint": "8.31.0",
        "eslint-config-next": "13.1.2",
        "typescript": "4.9.4"
    }
2. Delete node_modules folder
3. Delete package-lock.json (not sure why it is in my folder) and yarn.lock
4. yarn

Then the issue is gone.

**Reply by muxalko** [reply]
fixed for me

**Comment by lancejpollard** [comment]
I have this:

```tsx
/* eslint-disable @typescript-eslint/no-explicit-any */
import Document, {
  Html,
  Head,
  Main,
  NextScript,
  DocumentContext,
} from 'next/document'
import FONTS from '~/configurations/fonts'
import { ServerStyleSheet } from 'styled-components'

const fontsPaths = Object.keys(FONTS).map(name => FONTS[name][0])

export async function getInitialProps(ctx: DocumentContext) {
  const sheet = new ServerStyleSheet()
  const originalRenderPage = ctx.renderPage

  try {
    ctx.renderPage = () =>
      originalRenderPage({
        enhanceApp: App => (props: React.ComponentProps<any>) =>
          sheet.collectStyles(<App {...props} />),
      })

    const initialProps = await Document.getInitialProps(ctx)
    return {
      ...initialProps,
      styles: [initialProps.styles, sheet.getStyleElement()],
    }
  } finally {
    sheet.seal()
  }
}

export default function MyDocument() {
  return (
    <Html lang="en">
      <Head>
        {fontsPaths.map(font => (
          <link
            key={font}
            href={font}
            as="font"
            crossOrigin=""
          />
        ))}
      </Head>
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  )
}
```

And am getting this on `npm run build`:

```
info  - Linting and checking validity of types ...Failed to compile.

./pages/_document.tsx:38:8
Type error: 'Head' cannot be used as a JSX component.
  Its instance type 'Head' is not a valid JSX element.
    Type 'Head' is missing the following properties from type 'ElementClass': setState, forceUpdate, props, state, refs

  36 |   return (
  37 |     <Html lang="en">
> 38 |       <Head>
     |        ^
  39 |         {fontsPaths.map(font => (
  40 |           <link
  41 |             key={font}
$
```

Has anything else been figured out here?

**Reply by lancejpollard** [reply]
This goes away when using Next.js 14 and the app router system.

**Reply by bitcoinbrisbane** [reply]
> I have this:
> 
> ```tsx
> /* eslint-disable @typescript-eslint/no-explicit-any */
> import Document, {
>   Html,
>   Head,
>   Main,
>   NextScript,
>   DocumentContext,
> } from 'next/document'
> import FONTS from '~/configurations/fonts'
> import { ServerStyleSheet } from 'styled-components'
> 
> const fontsPaths = Object.keys(FONTS).map(name => FONTS[name][0])
> 
> export async function getInitialProps(ctx: DocumentContext) {
>   const sheet = new ServerStyleSheet()
>   const originalRenderPage = ctx.renderPage
> 
>   try {
>     ctx.renderPage = () =>
>       originalRenderPage({
>         enhanceApp: App => (props: React.ComponentProps<any>) =>
>           sheet.collectStyles(<App {...props} />),
>       })
> 
>     const initialProps = await Document.getInitialProps(ctx)
>     return {
>       ...initialProps,
>       styles: [initialProps.styles, sheet.getStyleElement()],
>     }
>   } finally {
>     sheet.seal()
>   }
> }
> 
> export default function MyDocument() {
>   return (
>     <Html lang="en">
>       <Head>
>         {fontsPaths.map(font => (
>           <link
>             key={font}
>             href={font}
>             as="font"
>             crossOrigin=""
>           />
>         ))}
>       </Head>
>       <body>
>         <Main />
>         <NextScript />
>       </body>
>     </Html>
>   )
> }
> ```
> 
> And am getting this on `npm run build`:
> 
> ```
> info  - Linting and checking validity of types ...Failed to compile.
> 
> ./pages/_document.tsx:38:8
> Type error: 'Head' cannot be used as a JSX component.
>   Its instance type 'Head' is not a valid JSX element.
>     Type 'Head' is missing the following properties from type 'ElementClass': setState, forceUpdate, props, state, refs
> 
>   36 |   return (
>   37 |     <Html lang="en">
> > 38 |       <Head>
>      |        ^
>   39 |         {fontsPaths.map(font => (
>   40 |           <link
>   41 |             key={font}
> $
> ```
> 
> Has anything else been figured out here?

As above, this fixed it for me

```bash
yarn add typescript @types/react @types/node -D
```

From package.json

```json
  "devDependencies": {
    "@types/node": "^20.14.8",
    "@types/react": "^18.3.3",
    "autoprefixer": "^10.4.13",
    "eslint": "^8.27.0",
    "eslint-config-next": "^13.0.3",
    "postcss": "^8.4.19",
    "tailwindcss": "^3.2.4",
    "typescript": "^5.5.2"
  }
```

**Comment by PaulRBerg** [comment]
[`syncpack`](https://github.com/JamieMason/syncpack) might be able to help with issues like this (related to accidentally using multiple `@types/react` versions).
